File: Digito Verificador/Modulo 11 Padrao/DV Gerador/gerardig_mod11.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int validarMatricula(const char* MATRICULA)
{
  if (strlen(MATRICULA) != 7)
    return 0;
  if (!isdigit(MATRICULA[0]) || !isdigit(MATRICULA[6]))
    return 0;
  if (MATRICULA[3] != '.')
    return 0;
  for (int I = 1; I <= 2; I++)
    if (!isdigit(MATRICULA[I]) || !isdigit(MATRICULA[I + 4]))
      return 0;
  return 1;
}

void removerCaractere(char CADEIA[], char CARACTERE)
{
  char *POS1 = CADEIA;
  char *POS2 = CADEIA;
  while (*POS1)
  {
    *POS2 = *POS1++;
    POS2 += (*POS2 != CARACTERE);
  }
  *POS2 = '\0';
}

char mod11(char MATRIC_CPT[])
{
  char MATRIC_SPT[8];
  int  MATRIC_INT[6];
  int  I, SOMA = 0;
  char DV;
  strcpy(MATRIC_SPT, MATRIC_CPT);
  removerCaractere(MATRIC_SPT, '.');
  for (I = 0; I <= 5; I++)
    MATRIC_INT[I] = MATRIC_SPT[I] - 48;
  for (I = 0; I <= 5; I++)
    SOMA += MATRIC_INT[I] * (7 - I);
  if ((SOMA % 11) == 0)
    DV = 48; // DV = 0
  if ((SOMA % 11) == 1)
    DV = 88; // DV = X
  if ((SOMA % 11) == 10)
    DV = 49; // DV = 1
  if ((11 - SOMA % 11) >= 2 && (11 - SOMA % 11) <= 9)
    DV = (11 - (SOMA % 11)) + 48;
  return DV;
}

int main(void)
{
  char MATRICULA[8], DV;

  do
  {
    printf("Informe matricula no formato 999.999: ");
    scanf("%s", MATRICULA);
    getchar(); // Lê o caractere de nova linha após a entrada
    if (!validarMatricula(MATRICULA))
    {
      printf("\n");
      printf("Matricula invalida. Tente novamente.\n");
      printf("\n\n");
    }
  } while (!validarMatricula(MATRICULA));

  DV = mod11(MATRICULA);
  printf("\n");
  printf("Matricula com DV = %s-%c\n", MATRICULA, DV);
  printf("\n");
  printf("Tecle <Enter> para encerrar... ");
  getchar();
  return 0;
}

File: Digito Verificador/Modulo 11 Padrao/DV Gerador/gerardig_mod11.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func validarMatricula(MATRICULA string) bool {
	if len(MATRICULA) != 7 {
		return false
	}
	if !isDigit(MATRICULA[0]) || !isDigit(MATRICULA[6]) {
		return false
	}
	if MATRICULA[3] != '.' {
		return false
	}
	for I := 1; I <= 2; I++ {
		if !isDigit(MATRICULA[I]) || !isDigit(MATRICULA[I + 4]) {
			return false
		}
	}
	return true
}

func isDigit(char byte) bool {
	return char >= '0' && char <= '9'
}

func removerCaractere(CADEIA string, CARACTERE string) string {
	return strings.ReplaceAll(CADEIA, CARACTERE, "")
}

func mod11(MATRIC_CPT string) string {
	MATRIC_SPT := MATRIC_CPT
	MATRIC_SPT = removerCaractere(MATRIC_SPT, ".")
	MATRIC_INT := make([]int, 6)
	for I := 0; I <= 5; I++ {
		MATRIC_INT[I], _ = strconv.Atoi(string(MATRIC_SPT[I]))
	}
	SOMA := 0
	for I := 0; I <= 5; I++ {
		SOMA += MATRIC_INT[I] * (7 - I)
	}
	var DV string
	if (SOMA % 11) == 0 {
		DV = "0"
	} else if (SOMA % 11) == 1 {
		DV = "X"
	} else if (SOMA % 11) == 10 {
		DV = "1"
	} else {
		DV = strconv.Itoa(11 - (SOMA % 11))
	}
	return DV
}

func main() {
	
	var MATRICULA string
	var DV string

	for {
		fmt.Print("Informe matricula no formato 999.999: ")
		fmt.Scanln(&MATRICULA)
		if !validarMatricula(MATRICULA) {
			fmt.Println()
			fmt.Println("Matricula invalida. Tente novamente.")
			fmt.Println()
		} else {
			break
		}
	}

	DV = mod11(MATRICULA)
	fmt.Println()
	fmt.Printf("Matricula com DV = %s-%s\n", MATRICULA, DV)
	fmt.Println()
	fmt.Print("Tecle <Enter> para encerrar... ")
	fmt.Scanln()
}
